### ISSUE TITLE
Can't restore deleted direct channel

### ISSUE THREAD

**Issue by andrewm41** [mattermost/mattermost-server#9001]
**registered user at oss.mattermost.com:**

curl -X POST -s -i -H 'Authorization: Bearer '"$TOKEN" -d '["USERNAME"]' https://oss.mattermost.com/api/v4/users/usernames

HTTP/1.1 200 OK
Server: nginx/1.4.6 (Ubuntu)
Date: Tue, 26 Jun 2018 11:20:02 GMT
Content-Type: application/json
Content-Length: 352
Connection: keep-alive
X-Request-Id: c544kaz6btnt5b74xo7duckw1y
X-Version-Id: 5.0.0.5.0.0.e98e0bd9f0deadf3b40e9593fb414748.false

[{"id":"epuggfymhjgktptq9cbaadoc5o","create_at":1529998666504,"update_at":1529998666504,"delete_at":0,"username":"USERNAME","auth_data":"","auth_service":"","email":"","nickname":"","first_name":"","last_name":"","position":"","roles":"system_user","locale":"en","timezone":{"automaticTimezone":"","manualTimezone":"","useAutomaticTimezone":"true"}}]

**created direct channel to myself:**
curl -X POST -s -i -H 'Authorization: Bearer '"$TOKEN" -d '["epuggfymhjgktptq9cbaadoc5o","epuggfymhjgktptq9cbaadoc5o"]' https://oss.mattermost.com/api/v4/channels/direct

**then deleted it:**
curl -X DELETE -s -i -H 'Author '"$TOKEN" https://oss.mattermost.com/api/v4/channels/xnd3whi7s3r1dkjcufbaeyg1me
it said OK

**now I can't create direct channel, because it was created already, and I can't restore it because it says:**
curl -X POST -s -i -H 'Author '"$TOKEN" https://oss.mattermost.com/api/v4/channels/xnd3whi7s3r1dkjcufbaeyg1me/restore
HTTP/1.1 401 Unauthorized
Server: nginx/1.4.6 (Ubuntu)
Date: Tue, 26 Jun 2018 11:25:51 GMT
Content-Type: application/json
Content-Length: 186
Connection: keep-alive
X-Request-Id: j1mazjhhiigjbkfd8gjs5f9rtw
X-Version-Id: 5.0.0.5.0.0.e98e0bd9f0deadf3b40e9593fb414748.false

{"id":"api.context.session_expired.app_error","message":"Invalid or expired session, please login again.","detailed_error":"","request_id":"j1mazjhhiigjbkfd8gjs5f9rtw","status_code":401}

**but my TOKEN is for sure valid, I can run any other requests with it**

Edited: there is a typo in authentication header, but bug is reproducible.

**Comment by amyblais** [comment]
Hi @andrewm41! Thank you for reporting this.

Would you happen to know if this issue occurs only at oss? Would you like to test on our [test server](https://rctesting.reddogsofwar.com/signup_user_complete/?id=7hwh373at38uunnzirao77psnh) to see if you see the same issue?

**Comment by andrewm41** [comment]
**== rctesting.reddogsofwar.com ==**
it is buggy but in different way

**Created user:**

curl -X POST -s -i -H 'Authorization: Bearer '"$TOKEN" -d '["USERNAME"]' https://rctesting.reddogsofwar.com/api/v4/users/usernames
HTTP/1.1 200 OK
Content-Type: application/json
X-Request-Id: ap3qqrw87ffqpk7ubntkqbps6c
X-Version-Id: 5.0.0.5.0.0.9318b5672b25b62162fe34ade6ad3fa0.true
Date: Wed, 27 Jun 2018 06:23:35 GMT
Content-Length: 345

[{"id":"xgynbjdmzpfhjmdrbxekmfg3fa","create_at":1530079537772,"update_at":1530079537871,"delete_at":0,"username":"amu","auth_data":"","auth_service":"","email":"","nickname":"","first_name":"","last_name":"","position":"","roles":"system_user","locale":"en","timezone":{"automaticTimezone":"","manualTimezone":"","useAutomaticTimezone":"true"}}]



**Created channel:**

curl -X POST -s -i -H 'Authorization: Bearer '"$TOKEN" -d '["xgynbjdmzpfhjmdrbxekmfg3fa","xgynbjdmzpfhjmdrbxekmfg3fa"]' https://rctesting.reddogsofwar.com/api/v4/channels/direct
HTTP/1.1 201 Created
Content-Type: application/json
X-Request-Id: baxruenbhtgoibwayqpon1mcor
X-Version-Id: 5.0.0.5.0.0.9318b5672b25b62162fe34ade6ad3fa0.true
Date: Wed, 27 Jun 2018 06:14:07 GMT
Content-Length: 322

{"id":"ys8fx6s1c3bw3xoqu7hqk5y13o","create_at":1530080044457,"update_at":1530080044457,"delete_at":0,"team_id":"","type":"D","display_name":"","name":"xgynbjdmzpfhjmdrbxekmfg3fa__xgynbjdmzpfhjmdrbxekmfg3fa","header":"","purpose":"","last_post_at":0,"total_msg_count":0,"extra_update_at":0,"creator_id":"","scheme_id":null}


**I can see channel:**

curl -X GET -s -i -H 'Authorization: Bearer '"$TOKEN"  "$MURI"/api/v4/channels/ys8fx6s1c3bw3xoqu7hqk5y13o/posts
HTTP/1.1 200 OK
Content-Type: application/json
Etag: 5.0.0.1530080427053
Expires: 0
X-Request-Id: 9sm6erg14jywdxaho78uiua66c
X-Version-Id: 5.0.0.5.0.0.9318b5672b25b62162fe34ade6ad3fa0.true
Date: Wed, 27 Jun 2018 06:20:27 GMT
Content-Length: 23

{"order":[],"posts":{}}


**But I can't delete it:**

curl -X DELETE -s -i -H 'Author '"$TOKEN" "$MURI"/api/v4/channels/ys8fx6s1c3bw3xoqu7hqk5y13o
HTTP/1.1 401 Unauthorized
Content-Type: application/json
X-Request-Id: at56udmumbnaxktqhgo7styh3r
X-Version-Id: 5.0.0.5.0.0.9318b5672b25b62162fe34ade6ad3fa0.true
Date: Wed, 27 Jun 2018 06:25:26 GMT
Content-Length: 198

{"id":"api.context.session_expired.app_error","message":"Invalid or expired session, please login again.","detailed_error":"UserRequired","request_id":"at56udmumbnaxktqhgo7styh3r","status_code":401}

**it says "Invalid or expired session" (code 401) but it is not**

**The situation is the same when I do these steps with cross user direct channel (using the second account).**

**Comment by amyblais** [comment]
Hi @andrewm41! Thanks for testing that out!

I have limited experience with API, so just wanted to confirm whether you are following our docs on [restoring a channel](https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels~1%7Bchannel_id%7D~1restore%2Fpost) and [deleting a channel](https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels~1%7Bchannel_id%7D%2Fdelete)

If yes, or my question doesn't make sense, just let me know and I'll ask an engineer to take a look.

**Comment by andrewm41** [comment]
Hello!
Yes, I used this api reference you mentioned above (https://api.mattermost.com/).

**Comment by amyblais** [comment]
Hi @andrewm41 - I checked with our engineers about this and they said it looks like your session token is invalid or expired (and that these are not bugs). 

Let us know any questions!

**Comment by jwilander** [comment]
@andrewm41 maybe it's not your token but some of your headers look wrong:

```
curl -X DELETE -s -i -H 'Author '"$TOKEN" 
```

That header should be `'Authorization: Bearer '"$TOKEN"`

**Comment by lieut-data** [comment]
I think I can reproduce this:
```
> http https://oss.mattermost.com/api/v4/users/me | jq '.id'
"k7665x8bw3fetfg5hni5mj5gdr"

> echo '["k7665x8bw3fetfg5hni5mj5gdr","k7665x8bw3fetfg5hni5mj5gdr"]' | http POST https://oss.mattermost.com/api/v4/channels/direct
HTTP/1.1 201 Created
Connection: keep-alive
Content-Length: 322
Content-Type: application/json
Date: Fri, 29 Jun 2018 18:44:09 GMT
Server: nginx/1.4.6 (Ubuntu)
X-Request-Id: 88zh3rjt9irybd3gocgzyxt8jo
X-Version-Id: 5.0.0.5.0.0.e98e0bd9f0deadf3b40e9593fb414748.false

{
    "create_at": 1530297849584,
    "creator_id": "",
    "delete_at": 0,
    "display_name": "",
    "extra_update_at": 0,
    "header": "",
    "id": "9shht9jw8trhd8uxowxwdd1x5w",
    "last_post_at": 0,
    "name": "k7665x8bw3fetfg5hni5mj5gdr__k7665x8bw3fetfg5hni5mj5gdr",
    "purpose": "",
    "scheme_id": null,
    "team_id": "",
    "total_msg_count": 0,
    "type": "D",
    "update_at": 1530297849584
}

> http DELETE https://oss.mattermost.com/api/v4/channels/9shht9jw8trhd8uxowxwdd1x5w
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 15
Content-Type: application/json
Date: Fri, 29 Jun 2018 18:44:35 GMT
Server: nginx/1.4.6 (Ubuntu)
X-Request-Id: qcz1dodrjbytbqk6syoqn649mr
X-Version-Id: 5.0.0.5.0.0.e98e0bd9f0deadf3b40e9593fb414748.false

{
    "status": "OK"
}

> echo '["k7665x8bw3fetfg5hni5mj5gdr","k7665x8bw3fetfg5hni5mj5gdr"]' | http POST https://oss.mattermost.com/api/v4/channels/direct
HTTP/1.1 400 Bad Request
Connection: keep-alive
Content-Length: 199
Content-Type: application/json
Date: Fri, 29 Jun 2018 18:44:45 GMT
Server: nginx/1.4.6 (Ubuntu)
X-Request-Id: 5hi3wc99rinxmnnfccr8638m4e
X-Version-Id: 5.0.0.5.0.0.e98e0bd9f0deadf3b40e9593fb414748.false

{
    "detailed_error": "",
    "id": "store.sql_channel.save_channel.previously.app_error",
    "message": "A channel with that URL was previously created",
    "request_id": "5hi3wc99rinxmnnfccr8638m4e",
    "status_code": 400
}

> http POST https://oss.mattermost.com/api/v4/channels/9shht9jw8trhd8uxowxwdd1x5w/restore
HTTP/1.1 403 Forbidden
Connection: keep-alive
Content-Length: 178
Content-Type: application/json
Date: Fri, 29 Jun 2018 18:45:06 GMT
Server: nginx/1.4.6 (Ubuntu)
X-Request-Id: n9oyqfymupntzdesh7jphfbfpe
X-Version-Id: 5.0.0.5.0.0.e98e0bd9f0deadf3b40e9593fb414748.false

{
    "detailed_error": "",
    "id": "api.context.permissions.app_error",
    "message": "You do not have the appropriate permissions",
    "request_id": "n9oyqfymupntzdesh7jphfbfpe",
    "status_code": 403
}
```

**Comment by amyblais** [comment]
Hi @andrewm41! I opened a ticket here for our engineers to investigate this: https://mattermost.atlassian.net/browse/MM-11118.

Thanks for finding this bug!

**Comment by andrewm41** [comment]
> maybe it's not your token but some of your headers look wrong:
> curl -X DELETE -s -i -H 'Author '"$TOKEN" 
> That header should be 'Authorization: Bearer '"$TOKEN"

yes, you are right, it was my typo from original post, I was bad in copy/paste from my terminal, and this typo didn't allow me to reproduce this bug on rctesting.reddogsofwar.com

thank you for noticing it

and with correct authentication token header I can reproduce it on rctesting.reddogsofwar.com:

**I deleted it successfully:**
curl -X DELETE -s -i -H 'Authorization: Bearer '"$TOKEN" https://rctesting.reddogsofwar.com/api/v4/channels/ys8fx6s1c3bw3xoqu7hqk5y13o
HTTP/1.1 200 OK
Content-Type: application/json
X-Request-Id: kw5uek3yupfpzm5ths7etzqggw
X-Version-Id: 5.0.0.5.0.0.9318b5672b25b62162fe34ade6ad3fa0.true
Date: Mon, 02 Jul 2018 05:30:16 GMT
Content-Length: 15

{"status":"OK"}

**and still I can't restore it:**
 curl -X POST -s -i -H 'Authorization: Bearer '"$TOKEN" https://rctesting.reddogsofwar.com/api/v4/channels/ys8fx6s1c3bw3xoqu7hqk5y13o/restore
HTTP/1.1 403 Forbidden
Content-Type: application/json
X-Request-Id: kn1x6cj4jjydiq4zbcp4aq96my
X-Version-Id: 5.0.0.5.0.0.9318b5672b25b62162fe34ade6ad3fa0.true
Date: Mon, 02 Jul 2018 05:35:56 GMT
Content-Length: 235

{"id":"api.context.permissions.app_error","message":"You do not have the appropriate permissions","detailed_error":"userId=xgynbjdmzpfhjmdrbxekmfg3fa, permission=manage_team","request_id":"kn1x6cj4jjydiq4zbcp4aq96my","status_code":403}

**Comment by lieut-data** [comment]
The underlying problem here was that we incorrectly allowed direct messages and group messages to be marked as deleted. This doesn't really make sense within our current data model, so the fix will be to prevent same.

Any incorrectly deleted direct or group messages will need to be manually patched via a database call:
{noformat}
UPDATE Channels SET DeleteAt = 0 WHERE Type IN ( 'D', 'G' );
{noformat}

**Comment by amyblais** [comment]
Closing this Issue as this has been fixed and final for v5.1 is cut today.
